libraries/AP_CANManager/AP_CANManager.h

```diff
@@ -64,7 +64,7 @@ class AP_CANManager
     // returns number of active CAN Drivers
     uint8_t get_num_drivers(void) const
     {
-        return _num_drivers;
+        return HAL_MAX_CAN_PROTOCOL_DRIVERS;
```

Comment:
I think there may be a problem with _driver_type_cache and this change. The common pattern is to loop over get_num_drivers() and check get_driver_type(). If that is the right type then we start de-referencing driver pointers. 
For this change to be safe I think we need to not set _driver_type_cache until after the init of an interface succeeds.
(ie. after can_initialised is true in AP_CANManager::init)

Reply:
@tridge We also check if the drivers returned is nullptr before dereferencing. If init has not been called the protocol driver pointer will NULL.

Reply:
I see your worry with this, also I think we will need to create an intermediate cache, so as to ensure the type is not changed by user while in init().